src/keys/bip39/mod.rs

```
@@ -0,0 +1,750 @@
+            Error::InvalidWord(word) => write!(f, "invalid word in mnemonic {}", word),
+            Error::InvalidEntropyLength(b) => write!(f, "invalid entropy length {}", b),
+            Error::InvalidChecksum(b) => write!(f, "invalid checksum length {}", b)
+        }
+    }
+}
+
+/// Type describing entropy length (aka word count) in the mnemonic
+pub enum WordCount {
+    /// 12 words mnemonic (128 bits entropy)
+    Words12 = 128,
+    /// 15 words mnemonic (160 bits entropy)
+    Words15 = 160,
+    /// 18 words mnemonic (192 bits entropy)
+    Words18 = 192,
+    /// 21 words mnemonic (224 bits entropy)
+    Words21 = 224,
+    /// 24 words mnemonic (256 bits entropy)
+    Words24 = 256,
+}
+
+/// A mnemonic is group of easy to remember words used in the generation of deterministic wallets
+pub struct Mnemonic {
+    /// The language this mnemoic belongs to
+    language: Language,
+    /// Vec of indices of words from the respective wordlist
+    words: Vec<u16>
+}
+
+impl Mnemonic {
+    /// Encode a mnemonic string to a `Mnemonic` type
+    pub fn parse_in(language: Language, mnemonic: &str) -> Result<Mnemonic, Error> {
+        let mnemonic: String = mnemonic.nfkd().collect();
+
+        let words: Vec<&str> = mnemonic.split_whitespace().collect();
+
+        if words.len() % 6 != 0 {
+            return Err(Error::InvalidWordCount(words.len()))
+        }
+
+        if words.len() < MIN_WORDS || words.len() > MAX_WORDS {
+            return Err(Error::InvalidWordCount(words.len()))
+        }
+
+        let mut indices: Vec<u16> = Vec::with_capacity(words.len());
+        let wordlist = language.wordlist();
+        for word in words {
+            let idx = match wordlist.iter().position(|&x| x == word) {
+                Some(i) => i as u16,
+                None => return Err(Error::InvalidWord(word.to_string()))
+            };
+
+            indices.push(idx);
+        }
```

After calculating the indices for every word, you may want  to recover the entropy, then calculate the checksum to compare it against the binary representation of mnemonic. In case, the calculated checksum does not match the checksum in the binary representation of the mnemonic, you will want to throw an error.

---

That's a good idea. Would you like to pick this up?

---

Sorry for the late reply. I will work on it